tools/bbc/bbc.cpp

```diff
@@ -138,33 +137,30 @@ void convertFortranSourceToMLIR(
   fir::KindMapping kindMap{&burnside.getMLIRContext()};
   burnside.lower(parseTree, nameUniquer);
   mlir::ModuleOp mlirModule = burnside.getModule();
-  mlir::PassManager pm = mlirModule.getContext();
 
-  pm.addPass(fir::createMemToRegPass());
-  pm.addPass(fir::createCSEPass());
-  pm.addPass(fir::createLowerToLoopPass());
-  pm.addPass(fir::createFIRToStdPass(kindMap));
-  pm.addPass(mlir::createLowerToCFGPass());
+  std::error_code ec;
+  llvm::ToolOutputFile out(OutputFilename, ec, llvm::sys::fs::OF_None);
+  if (ec) {
+    errs() << "can't open output file " + OutputFilename;
+    return;
+  }
 
   if (EmitLLVM) {
+    mlir::PassManager pm = mlirModule.getContext();
+    pm.addPass(fir::createMemToRegPass());
+    pm.addPass(fir::createCSEPass());
+    pm.addPass(fir::createLowerToLoopPass());
+    pm.addPass(fir::createFIRToStdPass(kindMap));
+    pm.addPass(mlir::createLowerToCFGPass());
     pm.addPass(fir::createFIRToLLVMPass(nameUniquer));
-    std::error_code ec;
-    llvm::ToolOutputFile out(OutputFilename + ".ll", ec,
-                             llvm::sys::fs::OF_None);
-    if (ec) {
-      errs() << "can't open output file " + OutputFilename + ".ll";
+
+    pm.addPass(fir::createLLVMDialectToLLVMPass(out.os()));
+    if (!mlir::succeeded(pm.run(mlirModule))) {
+      errs() << "oops, pass manager reported failure\n";
       return;
     }
-    pm.addPass(fir::createLLVMDialectToLLVMPass(out.os()));
-  }
-
-  if (mlir::succeeded(pm.run(mlirModule))) {
-    std::error_code ec;
-    raw_fd_ostream out(OutputFilename, ec);
-    if (!ec)
-      mlirModule.print(out);
   } else {
-    errs() << "oops, pass manager reported failure\n";
+    mlirModule.print(out.os());
```

Comment:
This looks a little off to me. The `-emit-llvm` flag is meant only to enable writing out the `file.ll` with the LLVM IR. It wasn't meant to enable and disable all the passes the pass manager runs. Each of those should have their own on/off option individually. Or, if we want to separate out "optimizations", they might be under a `-O` flag of some sort.  Getting both the LLVM IR and the MLIR dump as separate outputs from a single run of `bbc` seems desirable. Also, since the focus of `bbc` is testing Burnside, the LLVM IR output was to be opted into, but "lowering" should encompass more than just the bridge, I think. We want to run a few standard passes to make sure what the bridge is producing isn't garbage IR.

Reply:
Right, I made a small update adding a switch for llvm IR output filename and to control FIR passes independently of LLVM IR generation.
I am actually not sure what passes are required lowering steps and which one are "optimizations" (can be skipped to generate LLVM), so I use "--fir-passes" switch instead of "--optimize". Which passes are required if one then wants to lower to LLVM ?